Tracking issues page (#786)
sourcegraph/about: handbook/product/tracking_issues.md

> # Tracking issues

Reviewer: I am wondering if this should live under engineering instead of product.

If it is under product, it might feel like this is a process imposed on engineering by another function, but I don't think that is true. We do this because it is an effective way to communicate and plan engineering work to relevant stakeholders (the primary one being product).

I don't feel strongly about this but wanted to see what others thought: @christinaforney @lguychard

Reviewer 2: I think that's a great point to communicate. Where it lives matters less to me than that this is also written down. Just because a process is in engineering it could be mandated by another team. If instead we clearly communicate why we use the tool and how it serves the team, we resolve this concern.

Reviewer 3: > Where it lives matters less to me than that this is also written down.

+1

## Table of contents

1. [What is a Tracking issue?](#what-is-a-tracking-issue)
1. [Creating a Tracking issue](#creating-a-tracking-issue)
1. [Populating and maintaining a Tracking issue](#populating-and-maintaining-a-tracking-issue)
1. [Planning a milestone with a Tracking issue](#planning-a-milestone-with-a-tracking-issue)
1. [Using a Tracking issue for progress check-ins](#using-a-tracking-issue-for-progress-check-ins)
1. [Contributing to the Tracking issue tool](#contributing-to-the-tracking-issue-tool)

> ## What is a Tracking issue?

Reviewer: I don't think "Tracking" is a proper noun, so it doesn't need to be capitalized unless it is at the beginning of a sentence. Here and elsewhere

```suggestion
## What is a tracking issue?
```

> A Tracking issue is a GitHub issue that captures the planned and on-going work of a team's milestone, project, RFC, OKR or anything else of the sort. This artifact is a medium used for planning, progress check-ins and stakeholder communication.

Reviewer 4: ```suggestion
A Tracking issue is a GitHub issue that captures the planned and on-going work of a team's milestone, project, RFC, OKR or anything else that needs progress tracking. This artifact is a medium used for planning, progress check-ins and stakeholder communication.
```

You can take a look at [examples of open tracking issues](https://github.com/sourcegraph/sourcegraph/issues?q=is%3Aopen+is%3Aissue+label%3Atracking) to get a sense of what they look like.

> ## Creating a Tracking issue

Reviewer 2: ```suggestion
## Creating a tracking issue
```

> 1. Create a new GitHub issue with the [Tracking issue template](https://github.com/sourcegraph/sourcegraph/issues/new?assignees=&labels=tracking&template=tracking_issue.md&title=%24TEAM%3A+%24MILESTONE+Tracking+issue).

Reviewer: ```suggestion
1. Create a new GitHub issue with the [tracking issue template](https://github.com/sourcegraph/sourcegraph/issues/new?assignees=&labels=tracking&template=tracking_issue.md&title=%24TEAM%3A+%24MILESTONE+Tracking+issue).
```

> 1. Give it the **"$THING_BEING_TRACKED: Tracking Issue"** title.

Reviewer 2: Title case

```suggestion
1. Give it the **"$THING_BEING_TRACKED: Tracking issue"** title.
```

1. Assign it to the right milestone, creating such milestone if it doesn't yet exist.
1. Ensure the **tracking** label is set as well as the other labels you want to track. Issues and Pull Requests with those labels will show up in the tracking issue.

> 

Reviewer 5: ````suggestion
1. Add the following lines to the tracking issue where you want all the tracked issues and PRs to be displayed.
   ```
   <!-- BEGIN WORK -->
   <!-- END WORK -->
   ```

````

Reviewer: This is contained in the template so probably not necessary to add here?

> 1. Ensure the **tracking** label is set as well as the other labels you want to track. Issues and Pull Requests with those labels will show up in the tracking issue.

Reviewer 6: ```suggestion
1. Ensure the **tracking** label is set as well as the other labels you want to track. Issues and pull requests with those labels will show up in the tracking issue.
```

> ## Populating and maintaining a Tracking issue

Reviewer 2: ```suggestion
## Populating and maintaining a tracking issue
```

An open tracking issue is populated and kept up to date with the GitHub issues and pull requests labeled the same as the tracking issue (minus the `tracking` label). Optionally, a milestone can also be set. The tracking issue will be updated as these issues and pull requests are opened, changed, closed, merged, etc.

> This is done automatically by the [tracking-issue tool](#contributing-to-the-tracking-issue-tool) which [runs in response to GitHub issue events](https://github.com/sourcegraph/sourcegraph/blob/master/.github/workflows/tracking-issue.yml#L6) happening in [https://github.com/sourcegraph/sourcegraph](https://github.com/sourcegraph/sourcegraph).

Reviewer 6: In case I missed, if it's not already mentioned elsewhere. Even though the GitHub Action is deployed/defined/configured on our main repository, but we fetch properly labeled issues from all (or selected?) repositories under "sourcegraph" organization (e.g. sourcegraph/customer, sourcegraph/security).

Author: That's a good point that isn't mentioned yet. I'll write a few words.

### Rendering

There are a number of ways to affect how the workload section of a tracking issue is rendered.

#### Ordering

> The generated work load can be re-ordered manually in the GitHub UI and that ordering will be preserved when the tracking issue is updated.

Reviewer 3: ```suggestion
The generated workload can be re-ordered manually in the GitHub UI and that ordering will be preserved when the tracking issue is updated.
```

#### Labels

These are the standard labels you can use to refine how each work item in the tracking issue looks. You can add any number of these labels to an issue or pull request.

- **roadmap**: Issues related to work planned in our roadmap will be displayed the 🛠 emoji.

> - **customer**: Important issues reported or desired by a customer will be displayed with the 👩 emoji. If the issue body includes a Hubspot link to the customer's profile, then the emoji will be linked to it. If you have the Sourcegraph extension installed with *Experimental link previews* enabled and authorized the Hubspot's domain for it, then the customer's name will show up next to the emoji.

Reviewer 3: ```suggestion
- **customer**: Important issues reported or desired by a customer will be displayed with the 👩 emoji. If the issue body includes a Hubspot link to the customer's profile, then the emoji will be linked to it. If you have the Sourcegraph browser extension installed with *Experimental link previews* enabled and have authorized Hubspot's domain for it, then the customer's name will show up next to the emoji.
```

- **spike**: Time boxed investigation tasks meant to facilitate more granular planning will be displayed with the 🕵️ emoji.
- **bug**: An error, flaw or fault that produces an incorrect or unexpected result, or behavior. These will be displayed with the 🐛 emoji.

> - **debt**: Technical debt issues will be displayed with the 🧶emoji.

Reviewer: ```suggestion
- **debt**: Technical debt issues will be displayed with the 🧶 emoji.
```

- **estimate/$Nd**: A rough day level estimate of the task will be displayed alongside the emojis and summed up to a total workload in days that is displayed next to the teammate's GitHub handle.

> Apart from these standard labels, you can display any label inline (e.g. `perf`) by explicitly allowing it. This is done by adding a `<!-- LABEL: $YOUR_LABEL_NAME -->` comment to the tracking issue body outside the `<!-- BEGIN WORK-->` and `<-- END WORK -->` section which is reserved for the generated work loads.

Reviewer 3: ```suggestion
Apart from these standard labels, you can display any label inline (e.g. `perf`) by explicitly allowing it. This is done by adding a `<!-- LABEL: $YOUR_LABEL_NAME -->` comment to the tracking issue body outside the `<!-- BEGIN WORK-->` and `<-- END WORK -->` section which is reserved for the generated workloads.
```

> #### Pull Requests

Reviewer 6: ```suggestion
#### Pull requests
```

Pull requests, if labeled and milestoned correctly, will show up in their author's workload. They can be easily spotted by looking for the <img src="https://github.githubassets.com/images/icons/emoji/shipit.png" height=20 width=20/> emoji. Additionally, if a pull request's body contains the issue number (i.e. #1234) that it relates to, the it will show up under that issue as a nested list item.

> ## Planning a milestone with a Tracking issue

Reviewer: ```suggestion
## Planning a milestone with a tracking issue
```

> Planning a team's milestone happens in the last week of the previous milestone.

Reviewer 3: ```suggestion
Planning happens in the last week of the previous milestone.
```

> 1. [Open a new Tracking issue](#creating-a-tracking-issue) with the appropriate `team/$TEAM` label and the right milestone.

Reviewer: ```suggestion
1. [Open a new tracking issue](#creating-a-tracking-issue) with the appropriate `team/$TEAM` label and the right milestone.
```

1. Ask all members of the team through appropriate channels to create or assign issues to this milestone. When possible, issues should be estimated with the appropriate `estimate/$DAYS` label.

> 1. Once everyone's ready, have the tracking issue review by relevant parties (e.g. product team,).

Reviewer 3: ```suggestion
1. Once everyone's ready, have the tracking issue reviewed by relevant parties (e.g. product team, VP of engineering).
```

1. Make adjustments to the planned work with the team based on the feedback from the review.
1. Add the `planned/$MILESTONE` label to all planned issues. This is used to track what work was originally planned, even if it gets deprioritized during the milestone.

> All planned work items are meant to be finished in the iteration. If this doesn't happen, it is either due to unknown unknowns or lack of estimating accuracy. While we cannot work on the former, we can on the latter — over time, we aim to estimate better.

Reviewer 4: ```suggestion
All planned work items are meant to be finished in the iteration. If this doesn't happen, it is either due to unknown unknowns or poor estimating accuracy. While we cannot work on the former, we can improve the latter — over time, we aim to estimate better.
```

> All planned work items are meant to be finished in the iteration. If this doesn't happen, it is either due to unknown unknowns or lack of estimating accuracy. While we cannot work on the former, we can on the latter — over time, we aim to estimate better.

Reviewer 3: ```suggestion
All planned work items are meant to be finished in the iteration. If this doesn't happen, it is either due to unknown unknowns or inaccurate estimates. While we cannot work on the former, we can on the latter — over time, we aim to estimate better.
```

Teammates should be conservative by picking an estimate somewhere in between the median case and worst case scenarios, not best case. The higher someone's level of experience with the issue being estimated as well as estimation itself, the closer the estimate can be to the best case scenario. The opposite is also true.

Despite estimation having a bad reputation in some circles, we find it valuable to:

1. Prevent any single team member having more work assigned than available work days.
1. Ensure everyone has enough slack in the iteration (~20% of their time). This is to account for unexpected things, to foster creativity and to prevent burn-out.
1. Remind the person working on the issue to avoid time sinks or approaches that would miss the estimate.

Roadmap items may take more than an iteration to implement. As such please link to an issue which is scoped to this iteration. If it isn't clear how to break down a larger item in a scope increment, a **spike** of one or two days could be appropriate to determine such scope and then create the resulting issues for the iteration — it's fine to update the tracking issue with the results of a spike during the iteration.

> ## Using a Tracking issue for progress check-ins

Reviewer 2: ```suggestion
## Using a tracking issue for progress check-ins
```

> A Tracking issue can be used as a central communication hub for the progress being done around whatever is being tracked.

Reviewer 2: ```suggestion
A tracking issue can be used as a central communication hub for the progress being done around whatever is being tracked.
```

> To that end, those working on a tracking issue write high level periodic summaries of their progress for interested parties and other stake-holders to follow along without being exposed to too much detail. These summaries are written as comments on the tracking issue.

Reviewer 6: ```suggestion
To that end, those working on a tracking issue write high level periodic summaries of their progress for interested parties and other stakeholders to follow along without being exposed to too much detail. These summaries are written as comments on the tracking issue.
```

I don't know, but "stakeholders" is what we use [most of time in this repository](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/about%24+stakeholders&patternType=literal), and ["stake-holders" only has one occurrence](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/about%24+stake-holders&patternType=literal).

You can see an example of such a summary [here](https://github.com/sourcegraph/sourcegraph/issues/9910#issuecomment-616405856).

> ## Contributing to the Tracking issue tool

Reviewer 2: ```suggestion
## Contributing to the tracking issue tool
```
